arctic_training/trainer/dpo_trainer.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2025 Snowflake Inc.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict
+from typing import Union
+
+import torch
+
+from arctic_training.checkpoint.ds_engine import DSCheckpointEngine
+from arctic_training.checkpoint.hf_engine import HFCheckpointEngine
+from arctic_training.data.dpo_factory import DPODataFactory
+from arctic_training.model.hf_factory import HFModelFactory
+from arctic_training.model.liger_factory import LigerModelFactory
+from arctic_training.optimizer.adam_factory import FusedAdamOptimizerFactory
+from arctic_training.registry import register
+from arctic_training.scheduler.hf_factory import HFSchedulerFactory
+from arctic_training.tokenizer.hf_factory import HFTokenizerFactory
+from arctic_training.trainer import Trainer
+
+
+class DPOTrainerConfig:
+    ref_model: ModelConfig
+    """ Model configuration. """
+
+
+class DPOTrainer(Trainer):
+    name = "dpo"
+    config: DPOTrainerConfig
+    model_factory: Union[HFModelFactory, LigerModelFactory]
+    ref_model_factory: Union[HFModelFactory, LigerModelFactory]
+    checkpoint_engine: Union[DSCheckpointEngine, HFCheckpointEngine]
+    optimizer_factory: Union[FusedAdamOptimizerFactory]
+    scheduler_factory: Union[HFSchedulerFactory]
+    tokenizer_factory: Union[HFTokenizerFactory]
```

Comment:
```suggestion
    optimizer_factory: FusedAdamOptimizerFactory
    scheduler_factory: HFSchedulerFactory
    tokenizer_factory: HFTokenizerFactory
```

No need for union, we support both single type hints and unions now!